daffodil-io/src/main/scala/org/apache/daffodil/io/DirectOrBufferedDataOutputStream.scala

```diff
@@ -961,7 +1030,12 @@ object DirectOrBufferedDataOutputStream {
    * Factory for creating new ones/
    * Passing creator as null indicates no other stream created this one.
    */
-  def apply(jos: java.io.OutputStream, creator: DirectOrBufferedDataOutputStream, isLayer: Boolean = false) = {
+  def apply(
+    jos: java.io.OutputStream,
+    creator: DirectOrBufferedDataOutputStream,
+    isLayer: Boolean = false,
+    chunkSizeInBytes: Long = 4096,
```

Comment:
Tiny size for modern I/O. 64K minimum is what I'd suggest. 
Shouldn't this default come from the tunable now?

Reply:
I think 4096 was taken from the blob code I wrote. And I think I used 4096 because thats the size of a page on modern kernels. So i have no idea if it was reasonable. Though 64K feels a little big to me. That's a decent chunk of memory that Java is going to need to allocate. For comparison, the JDK Files.copy (which I think is similar to what we're doing?) uses a array of size BUFFER_SIZE:

https://github.com/openjdk-mirror/jdk/blob/jdk8u/jdk8u/master/src/share/classes/java/nio/file/Files.java#L2902

which is initialized to 8192:

https://github.com/openjdk-mirror/jdk/blob/jdk8u/jdk8u/master/src/share/classes/java/nio/file/Files.java#L2752

So maybe we should bump it up? Might be interesting to do a quick performance test to see if there's a big differentce bettween 4k, 8k, and 64k.

Though, I guess in theory this chunkSize is only ever going to be used when there are huge files (either BLOBS or large buffers) so maybe an allocation of 64k is going to miniscule compared to reading 2G+ of data?

Reply:
Just sayin',  these sizes, like 4K, are the same buffer sizes that were used when I was in college in 1979 using pdp11 running at 8 Megahertz. 

This is sequential I/O, not random.  No compromises for storing small fragments of data inefficiently are needed. 

With CPUs and I/O systems being nearly 3 decimal orders of magnitude faster, I think megabytes are the new Kilobytes.   

Of course perf testing should be the answer, but I would not bother to performance test yet. There is so much baggage in the unparser currently (due to optimizations that are supposed to happen, but which do not take place) that you'd never see anything pop up. 

See: https://issues.apache.org/jira/browse/DAFFODIL-2238. 

We are executing tons of excess overhead, capturing the content and value length for every element, winding into and out of fillBytes unparsers that do nothing, etc. These things need to get fixed first. Just print out the unparser for a small simple binary schema. It's scary huge.